feat(resourcemanager): replace `Scope` with `ArmAudience`

archerzz opened the pull request:
1. create an unsealed enum `ArmAudience` which represents Azure cloud audience (e.g. account types)
2. change `ArmClientOptions.Scope` to `ArmClientOptions.Audience`
3. in `ArmClient`, compose the scope from audience

resolve #26635

# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/main/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

archerzz committed:
feat(resourcemanager): replace `Scope` with `ArmAudience`

1. create an unsealed enum `ArmAudience` which represents Azure cloud audience (e.g. account types)
2. change `ArmClientOptions.Scope` to `ArmClientOptions.Audience`
3. in `ArmClient`, compose the scope from audience

resolve #26635

m-nash commented on sdk/resourcemanager/Azure.ResourceManager/src/ArmAudience.cs:
```diff
@@ -0,0 +1,61 @@
+    /// <summary>
+    /// ArmAudience represents the audience of Azure Cloud.
+    /// </summary>
+    public class ArmAudience : IEquatable<ArmAudience>
```
Can we make this a readonly struct?

archerzz replied:
done

m-nash commented on sdk/resourcemanager/Azure.ResourceManager/src/ArmAudience.cs:
```diff
@@ -0,0 +1,61 @@
+        public override bool Equals(object obj) => obj is ArmAudience other && Equals(other);
+
+        /// <inheritdoc />
+        public bool Equals(ArmAudience other) => string.Equals(_value, other._value, StringComparison.InvariantCultureIgnoreCase);
```
use OridinalIngoreCase

schaabs replied:
Given that these are urls should we be doing case sensitive comparisons rather than ignoring case?

m-nash replied:
oh yeah good point, use `Ordinal` here

archerzz replied:
done

m-nash commented on sdk/resourcemanager/Azure.ResourceManager/src/ArmClient.cs:
```diff
@@ -96,6 +96,11 @@ public ArmClient(TokenCredential credential, string defaultSubscriptionId, Uri b
                 new Subscription(this, Subscription.CreateResourceIdentifier(defaultSubscriptionId));
         }
 
+        private static string GetAuthenticationScope(ArmAudience audience)
```
Should this be a method on ArmAudience itself?

archerzz replied:
I add `ArmAudience.DefaultScope` property. I also add `ArmAudience.GetScope` method, but keep it private since I'm not sure if there is any use case that people want to compose other kinds of scope.

m-nash commented on sdk/resourcemanager/Azure.ResourceManager/src/ArmAudience.cs:
```diff
@@ -0,0 +1,61 @@
+    {
+        // name after the `name` property of returned audience from https://management.azure.com/metadata/endpoints?api-version=2019-11-01
+        /// <summary> Azure Public Cloud. </summary>
+        public static readonly ArmAudience AzureCloud = new ArmAudience("https://management.azure.com/");
```
Does this now double as the Uri as well?  Should we remove those params from the ArmClient ctor and use this from the options?

schaabs replied:
Hmm... this is a good point. If the audience is always matches the base URI provided when the client is constructed the `Audience` option might not be needed at all. We've only been adding this option for service clients where we can't determine the audience from the service endpoint provided in client construction. For instance services which support custom domain names. If this isn't the case for ARM we might be able to remove the option all together.

archerzz replied:
There are two issues we need to consider.

1. From the [response of this API](https://management.azure.com/metadata/endpoints?api-version=2019-11-01), there could be multiple audience per endpoint. Will the audience grow? If so, then we cannot use its value as the base URI.
2. Audience has [a trailing slash issue](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent#trailing-slash-and-default). For audience, we need to keep the extra slash, but for base URI we should not keep it.

archerzz replied:
Take Java SDK as reference. They bundle the environment information altogether in [AzureEnvironment](https://github.com/Azure/azure-sdk-for-java/blob/2cd30e97440042b17a94db3438865ed530fb4c08/sdk/core/azure-core-management/src/main/java/com/azure/core/management/AzureEnvironment.java).

- `managementEndpointUrl` is actually audience. I get confirmation that they never use it as a URL.
- `resourceManagerEndpointUrl` is the mgmt endpoint URL.

Besides, they allow users to specify their own scope in https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/resourcemanager/azure-resourcemanager-resources/src/main/java/com/azure/resourcemanager/resources/fluentcore/arm/AzureConfigurable.java#L80

m-nash replied:
Lets follow the Java pattern to keep customers from being able to specify incompatible values.

archerzz committed:
update according to review comments

1. change `ArmAudience` type to `readonly struct`
2. use `StringComparision.Ordinal` when compare audience value
3. add `DefaultScope` property to `ArmAudience`, and use its value when composing `BearerTokenAuthenticationPolicy`

m-nash commented on sdk/resourcemanager/Azure.ResourceManager/src/ArmClient.cs:
```diff
@@ -79,11 +79,11 @@ public ArmClient(TokenCredential credential, string defaultSubscriptionId, Uri b
 
```
https://apiview.dev/Assemblies/Review/f6d0e5c39961448fb07909c216860ffc#de662a76a3b14b1d8b125104d902a08d

archerzz committed:
consolidate baseUri and audience

1. use `ArmEnvironment` to consolidate both base uri and audience
2. refactor `ArmOptions` to include `ArmEnvironment`
3. refactor `ArmClient` constructors to adopt base uri and audience in `ArmOptions`, remove `baseUri` parameter

archerzz committed:
remove ctor(cred, options)

archerzz committed:
update api

archerzz committed:
fix broken test

archerzz commented:
major changes since last review:
- Delete `ArmAudience` and add `ArmEnvironment` to include both base uri and audience
- Remove `baseUri` parameter from constructor of `ArmClient`. Now we fetch the base uri from `ArmOptions.Environment.BaseUri`
- Clean up constructors of `ArmClient`

m-nash commented on sdk/resourcemanager/Azure.ResourceManager/src/ArmClient.cs:
```diff
@@ -42,50 +38,28 @@ protected ArmClient()
-        /// <param name="options"> The client parameters to use in these operations. </param>
-        /// <exception cref="ArgumentNullException"> If <see cref="TokenCredential"/> is null. </exception>
-        public ArmClient(TokenCredential credential, string defaultSubscriptionId, Uri baseUri, ArmClientOptions options = default)
+        public ArmClient(TokenCredential credential, string defaultSubscriptionId = default, ArmClientOptions options = default)
```
This is the most concise way to code the parameter overloads, but the feedback from customers / archboard is that we want to have non optional ctors.  We should have exactly these 3 overloads here.

ArmClient(TokenCredential credential)
ArmClient(TokenCredential credential, string defaultSubscriptionId)
ArmClient(TokenCredential credential, string defaultSubscriptionId, ArmClientOptions options)

archerzz replied:
done

archerzz committed:
update according to the review comments

1. remove null check in `ArmOptions.Environment`, since it's non-nullable
2. make default value of `ArmEnvironment` as `ArmEnvironment.Cloud`
3 update constructors of `ArmOptions`, split the optional parameter version into 3 versions with explicit paramters

archerzz commented on sdk/resourcemanager/Azure.ResourceManager/src/ArmClientOptions.cs:
```diff
@@ -15,10 +16,21 @@ public sealed class ArmClientOptions : ClientOptions
+            get => _environment;
+            set
+            {
+                if (value == null)
```
I realize this property is non-nullable, so the null check here is not necessary. However, I find one issue in `ArmEnvironment`: the default value is carries empty properties. I changed the `default(ArmEnvironment)` to be `ArmEnvironment.AzureCloud`, and add tests accordingly.

archerzz committed:
add missing api change

archerzz committed:
simplify the `ArmEnvironment` by removing default value

archerzz committed:
remove excessive null check

m-nash commented on sdk/resourcemanager/Azure.ResourceManager/src/ArmEnvironment.cs:
```diff
@@ -0,0 +1,89 @@
+    /// <summary>
+    /// ArmEnvrionment represents the information of an Azure Cloud environment.
+    /// </summary>
+    public readonly struct ArmEnvironment : IEquatable<ArmEnvironment>
```
sorry one final thing, make this sealed
